frontend/src/modules/labwork/services.ts

```diff
@@ -154,4 +173,99 @@ export function findChangedStepsInSummary(oldSummary: LabworkSummary, newSummary
+	// we gather information about where in each workflow a protocol is used. The map key is
+	// the protocol id, and one entry is created for each step that is found using that protocol.
+	// It's gnarly code! Sorry, but my first implementation had to traverse the workflows several times.
+	type ProtocolEntry = {
+		workflowID: FMSId
+		stepOrder: number
+		numSteps: number
+		step: WorkflowStep
+	}
+
+	const workflowOccurences = new Map<FMSId, ProtocolEntry[]>()
+
+	workflows.forEach(workflow => {
+		workflow.steps_order.forEach(step => {
+			const entry : ProtocolEntry = {
+				workflowID: workflow.id,
+				stepOrder: step.order,
+				numSteps: workflow.steps_order.length,
+				step: step
+			}
+			const list = workflowOccurences.get(step.protocol_id)
+			if (list) {
+				list.push(entry)
+			} else {
+				workflowOccurences.set(step.protocol_id, [entry])
+			}
+		})
+	})
+
+	
+	function compareProtocolPositionsInWorkflows(protocolID_A: FMSId, protocolID_B: FMSId) {
+		const entriesA = workflowOccurences.get(protocolID_A)
+		const entriesB = workflowOccurences.get(protocolID_B)
+
+		let numBefore = 0
+		let numAfter = 0
+
+		// Compare the position of protocol A with the position of protocol B in each
+		// workflow that contains the two. Count the number of times that A appears after
+		// B in the workflow, and the number of time A appears before B. Use that to decide
+		// if A is normally before or after B in the sort order.
+		if (entriesA && entriesB) {
+			for(const entryA of entriesA) {
+				// The same protocol can appear multiple times, so we have to check each instance
+				// of B in the workflow
+				const allBEntriesInWorkflow = entriesB.filter(entry => entry.workflowID === entryA.workflowID)
+				allBEntriesInWorkflow.forEach(entryB => {
+					if (entryA.stepOrder < entryB.stepOrder) {
+						numBefore++
+					} else {
+						numAfter++
+					}
+				})
+			}
+
+			// If numBefore and numberAfter are both zero then there was no case where both A and B
+			// are used in the same workflow, so we can't compare their positions in the workflow.
+			// Instead, compute an average "distance" for A and B. This is the step order divided by
+			// the number of steps in the workflow, averaged over all workflows the protocol appears in.
+			// The idea is to figure out if the protocol appears near the beginning of workflows or near
+			// the end. Without this, rarely used protocols tend to end up at the top of the list 
+			// (eg DBNSEQ Preparation).
+			if (numBefore === 0 && numAfter === 0 && entriesA.length > 0 && entriesB.length > 0) {
+				const distanceA = entriesA.reduce((acc, entry) => {
+					return acc + (entry.stepOrder / entry.numSteps)
+				}, 0) / entriesA.length
+				const distanceB = entriesB.reduce((acc, entry) => {
+					return acc + (entry.stepOrder / entry.numSteps)
+				}, 0) / entriesB.length
+
+				return distanceA - distanceB
+			} else {
+				return numAfter - numBefore
```

Comment:
It's likely going to give wonky results as we are not using a common metric for all steps... Both metrics are not on the same scale... This seem to be working for now but it's likely to need rework down the line. Let's roll with it, but ordering usually use only a single criterion (or a weighted combination).

Reply:
Yes, I'm sure that there is a better algorithm - this is a bit hacky, I agree, but it got a decent result for now.